Yaxis bug fix and beautification

Yida-Lin opened the pull request:
1. YAxis widget was terribly out of sync when doing channel overlapping, and there were a ton of mistakes in the code.

2. Redesigned the adapt window widget.

cbrnr commented:
Thanks @Yida-Lin, I'll take a look. It's getting a bit difficult to keep track of all changes - could you use _one_ PR per feature change/bug fix please?

Yida-Lin commented:
@cbrnr Sure, #35 actually overwrites #33. We can close #33 but preserve the conversation.

This PR is much more substantial than #33.

cbrnr commented:
OK, #33 was a bit weird anyway because it was your master :-). Can you fix those warnings in this one?

Yida-Lin commented:
@cbrnr Sure will try :)

cbrnr commented:
I like the new widget, but I've already thought about moving most of the stuff into a settings dialog because it's not really necessary to have this fourth mode in SigViewer. Also, right now the widget consumes a lot of vertical space and some labels are not readable:

<img width="1441" alt="screen shot 2017-02-20 at 09 13 59" src="https://cloud.githubusercontent.com/assets/4377312/23116935/823fc564-f74d-11e6-8867-79c33c01edaa.png">

I suggest to separate the widget redesign from the rest of this PR - could you please create a separate PR for this?

Yida-Lin commented:
@cbrnr Yes, in any case we should think of a way to preserve those bars or the knob if possible, while cutting the unnecessary mode---- I like those interactive stuff :) , especially the overlapping one. If the knob takes too much space a horizontal slider will do the job.

cbrnr commented:
Yes, let's think about that some more. For this PR, I thought we could remove those black vertical lines:

![screen shot 2017-02-20 at 11 46 12](https://cloud.githubusercontent.com/assets/4377312/23121907/8190a3a8-f762-11e6-9db3-ebe0974d50b1.png)

They distract from the signals, and the one at t = 0 is not visible when you scroll further to the right anyway.

cbrnr commented:
The x-grid is now synced to the x ticks on the x axis. Therefore, I suggest that we draw the grid again (in the same light color as the y grid).

Only thing I didn't find so far: for XDF files, the x grid is disabled - where is this setting?

cbrnr commented:
Also, we need to add a GUI option to enable/disable the x and y grids.

cbrnr commented:
And, toggling the x axis does not update the view - so that the last channel is only shown correctly if the x axis is disabled (because the x axis widget partly overlaps the last channel).

cbrnr commented:
Here's a to do list for this PR:

- [ ] Enable x-grid also for XDF files by default
- [ ] Add GUI options to enable/disable x- and y-grids
- [ ] Update view if x-axis widget is toggled (last channel needs to be updated)

Yida-Lin commented:
@cbrnr Thank you Clemens. However, I am not sure the result is very satisfying, yet: I don't find the parameters on the left clear enough to denote the signals.

Also, now the entire thing is a flat plain white panel, with no boundaries between channels at all. I am not sure whether this is easy to read. 

![image](https://cloud.githubusercontent.com/assets/20526646/23150090/a148974e-f7be-11e6-80fe-805c30a87270.png)
![image](https://cloud.githubusercontent.com/assets/20526646/23150165/6a5ab3c4-f7bf-11e6-8eb2-6c1c0f594fd9.png)

If you prefer removing the black boundaries, I propose we should consider some sort of background color----doesn't have to be as deep and aggressive as the Digital Performer (a music editing software), but maybe some light shades.

![image](https://cloud.githubusercontent.com/assets/20526646/23150119/0201fc10-f7bf-11e6-87b2-c7eb8b26fde2.png)

Yida-Lin commented:
@cbrnr Also, if new GUI options were added, may I redesign the layout?

Yida-Lin commented:
@cbrnr Hi Clemens!

1. >Enable x-grid also for XDF files by default.

This is now done.

2. >Add GUI options to enable/disable x- and y-grids

Also done.

3. >Update view if x-axis widget is toggled (last channel needs to be updated)

Not sure what happens on other OS, but on Windows I don't see the x-axis widget interferes with the last channel in any way. 

![image](https://cloud.githubusercontent.com/assets/20526646/23195776/571dd650-f883-11e6-8af1-aa6ae34752e2.png)

![image](https://cloud.githubusercontent.com/assets/20526646/23195790/5fa8e314-f883-11e6-9fc3-b1b9fbe09b36.png)


4. >Draw event rectangles only in signal regions when channels overlap (see screenshot below)

Now finished.

cbrnr commented:
Thanks @Yida-Lin, a lot of great new stuff :+1: and some open questions ❓ :

1. Here's how to reproduce the issue with toggling the x-axis. Open a file (let's use `sample_nan.gdf` here) and switch to View Options mode. Everything looks fine, the x-axis is shown. Then disable the x-axis. Everything still looks fine. However, if you then re-enable the x-axis, it partially overlaps the last channel (see screenshot).

![screen shot 2017-02-22 at 09 38 32](https://cloud.githubusercontent.com/assets/4377312/23203426/bda33050-f8e2-11e6-9076-0bb257a142b5.png)

2. There seems to be a problem with the y-axis widget. The units are sometimes not shown at all for some channels - see the screenshot from you above, this happens for the first channel in `sample_nan.gdf`. It also happens for any other file, including XDF files. I think a value close to the channel mean should always be shown.

3. Regarding the visual style: all EEG visualizations that I know do not separate channels by lines. Therefore, people will be more used to this kind of style. Plus, I like it better (less visual clutter). However, nothing is stopping us from adding an option to show/hide channel separators (but the default should be hide). I also think that you're right, currently the y-axis widget does not display enough information to make the channels easily separable - but see issue 2 above, if this is resolved I don't think we even need channel separators.

4. I like the new View Options widget, but it needs some minor adaptations. I'll push a commit with some changes soon.

5. I didn't remove the text edit widget on purpose - sorry about that, can you please add it back?

cbrnr commented:
Maybe we should refactor the 2 actions (zero line centered/zero line fitted) into one action (toggle channel centering).

Yida-Lin commented:
@cbrnr Hi Clemens!

I made some further improvements and commented diligently in some confusing parts. 

I am sorry that I couldn't completely agree on the approach of taking boarders all away; to me the numbers on the left are cluttering and super confusing.

![image](https://cloud.githubusercontent.com/assets/20526646/23246102/4cd955f2-f95f-11e6-8561-0dadad67df49.png)

I added part of the lines back. I thought it looked better, and easier to understand. 

![image](https://cloud.githubusercontent.com/assets/20526646/23246389/2ab7530a-f961-11e6-9b7c-688b7fca6b4c.png)

I didn't have time to finish other problem just yet. But please help me think how we can setup an online repository, that whenever we release new versions in the future, the maintenance tool (provided by Qt Installer Framework, in the root directory of the new SigViewer) can automatically update SigViewer.

The details can be found [here](http://doc.qt.io/qtinstallerframework/ifw-updates.html).

I thought this is super critical. So long as we have this feature ready, we can take time to make further improvements in a slow but steady pace, and push the updates to the users whenever new features are available.

cbrnr commented:
OK, so why don't we make these horizontal lines an option? Here's what I suggest: these lines should separate the channels, so we don't need a line above the first and below the last channel. If the lines are turned on, they should extend from the y-axis widget over the channel view to the label widget (that is, the lines go completely from left to right if they are enabled). We could add a checkbox to the Channel Overlap section, because the separating lines can only be shown if there is no overlap!

cbrnr commented:
Regarding the installer, do you mean we should add a check for updates feature? This would be super nice, but can we work on this after the 0.6 release?

cbrnr commented:
Also, I'd argue for not including keyboard shortcuts in tool tips, I think it's sufficient if they are shown in the menu.

Yida-Lin commented:
@cbrnr No, the updates feature is nice, but it's not something we can come up in one day or two.

What I am talking about is this: which can be done soon if we figure out how the [remote repository](http://doc.qt.io/qtinstallerframework/ifw-updates.html) works.

![2017-02-23 3 _li](https://cloud.githubusercontent.com/assets/20526646/23249608/3d8133c2-f973-11e6-9d54-bc5cc0aba584.jpg)

cbrnr commented:
Hm, I'm not sure if this is important for this release. We'd have to look into some questions, such as is this available across platforms? Do we even want that? Personally, I'd prefer a check for updates menu entry, and if there's a newer version available a dialog should pop up with a notification and a download link. In addition, this would not work nicely on a Linux system, which uses its own native package manager for updates.

Yida-Lin commented:
@cbrnr I thought it would be the most ideal when an application is smart enough to update itself without even manually clicking on some link or downloading it. This can't be done without some sort of framework.

cbrnr commented:
I'm not a big fan of automatic updates. Yes, notifications are useful, but I'd leave it up to the user to perform the actual update or not. There are use cases where a user might want to stay with a specific version during the course of a project.

Yida-Lin commented:
@cbrnr Ok that's cool, otherwise I think this PR has been mostly finished; maybe with only the x widget issue left, which I will take a look tomorrow.

cbrnr commented:
Thanks! Here are the remaining issues for this PR:

- [ ] The x-axis widget overplots the last channel when toggling.
- [ ] The values on the y-axis widget are sometimes missing.
- [ ] Lines separating the channels should be optional with a checkbox (default off).
- [ ] Missing text edit widget for the New Event mode.
- [ ] Remove warnings (see below).

```
src/gui/signal_view_settings.cpp:17:7: warning: field 'enable_x_grid_' will be initialized after field 'enable_y_grid_' [-Wreorder]
      enable_x_grid_ (true),
      ^
src/gui_impl/signal_browser/signal_browser_model_4.cpp:56:3: warning: field 'initialized_' will be initialized after field 'show_x_grid_' [-Wreorder]
  initialized_ (false),
  ^
src/gui_impl/signal_browser/signal_browser_model_4.cpp:57:3: warning: field 'show_x_grid_' will be initialized after field 'show_y_grid_' [-Wreorder]
  show_x_grid_ (true),
  ^
```

Yida-Lin commented:
@cbrnr Hi Clemens!

> The x-axis widget overplots the last channel when toggling.

I am looking at the code but haven't diagosed the problem yet. In the meantime, I found that, if you keep toggling the x widget (let's say 20 times), you will find the signals are getting narrower and narrower... weird.

> The values on the y-axis widget are sometimes missing.

Fixed.

> Lines separating the channels should be optional with a checkbox (default off).

Need some more time to work on it...

> Missing text edit widget for the New Event mode.

Actually that was a mistake. Earlier I was opening a GDF file and I forgot about that when I was designing the widget I set it to XDF only, because somehow I thought it doesn't work in GDF files. So currently it still works in XDF files. I added a messagebox warning that customized events might not be properly imported...

> Remove shortcuts from tool tips.

I commented it out. However, just sharing as personal experience, I never noticed there were shortcuts despite having been using SigViewer everyday for half a year, until I saw those short cuts in the source code...

> Remove warnings (see below).

Fixed.

cbrnr commented on src/gui/signal_view_settings.cpp:
```diff
@@ -15,7 +15,8 @@ SignalViewSettings::SignalViewSettings (ChannelManager const& channel_manager)
       enable_x_grid_ (true),
-      enable_y_grid_ (true)
+      enable_y_grid_ (true),
+      enable_boarder_lines (true)
```
Just a minor thing, it is spelled "border" :-)

Yida-Lin replied:
LOL... Can't believe this :) 

I am sorry I need to go to bed early tonight... Meeting Paul tomorrow morning. Will keeping working on it tomorrow :)

cbrnr commented:
Here's an idea how we could summarize some features in the View Options widget. We could create a group called "Channels" and put the following options in there: Overlap, Center Channels, and Border lines (note I would call this option Channel Separators, also in the source code). Channel Separators are automatically disabled if there is some overlap.

cbrnr commented:
Regarding the channel separators, you said that right now only the rectangle around a channel is working. Could we please only add horizontal lines and not a rectangle? I don't like the vertical line at time zero at all (plus it is inconsistent anyway, because there is no vertical line at the end of the signal).

cbrnr commented:
And I found a new issue: When you toggle the x- and y-grids (let's say you turn it off), and then you zoom into the channels by decreasing the Channels value in the Page Settings group, the grids will reappear and the checkboxes are not in sync anymore.

Yida-Lin commented:
I believe the separator lines are fixed now. 

Still trying to fix the x axis widget but haven't had much progress yet...

Yida-Lin commented:
@cbrnr X Widget overlapping problem fixed. Phew, finally.

Yida-Lin commented:
@cbrnr Currently most of the features in this PR have been finished.

However, I am thinking of another possibility: the _libxdf_ can be conveniently adopted by other applications. However, in collaboration with SigViewer, it's not perfect: a simple example being that it evades using any Qt functions. Thus, when loading data, it first loads everything into a STL container, later passed to SigViewer and SigViewer copy it to QVectors. This is unnecessary and time inefficient: what if we keep libxdf available solely for other applications, but for SigViewer, we integrate the code directly instead of using it as an `.a` file, this way I can use lot of Qt Functions and avoid redundant copying which can radically optimize performance. If we can load a 600MB file in 3 seconds, how sweet would that be. Or, I can tailor a special Qt edition of libxdf exclusively for SigViewer. Personally I prefer the former option. Not sure what you think.

Another option is we avoid using QVector in SigViewer but use std::vector in SigViewer instead. But this involves some internal changes in SigViewer.

Yida-Lin commented:
@cbrnr Sorry I was too bored so I added some more unnecessary stuff... feel free to revert it :) 

![image](https://cloud.githubusercontent.com/assets/20526646/23351843/a92d25ce-fc91-11e6-9733-6b426ef11ffe.png)

cbrnr commented:
@Yida-Lin, thanks for all the new stuff you added! Let's try to get 0.6 ready ASAP, so here are my comments regarding this PR:

1. Grouping items in the View Options widget is a good idea, but there shouldn't be too many groups. I'd still prefer a Channels group with the following items in it: (a) Center around mean, (b) Separators, and (c) Overlap.
2. I think adding an option to customize the x- and y-grids is useful, but there's no need to change the label colors. I'd also rather not have an option to change the color of the channel separators. Let's leave those black for now.
3. With all this customization going on, I think we should add options to revert to the default values. Either on a per-setting basis (which means we would have to add such an option for each setting), or globally.
4. Borderline reminds me of a psychological condition :smile:, so if you don't mind let's rename these lines to separators. To avoid long names, we don't have to call them ChannelSeparators, so the function you mentioned could be called `enableSeparators()`.
5. The separators should be drawn on top of the x-and y-grids. Currently, the grids overplot the separators.
6. The small problem you mention should not happen at all, or at least the way it works now is flawed. For example, zoom out of a signal completely. Then, SigViewer will show the complete signal using all available space, and the x-axis scrollbar is disabled. However, if you then change the window size (i.e. make it smaller in the x-direction), parts of the signal disappears, but the x-scrollbar is still disabled. Not sure how to solve this, do you have a suggestion?
7. Another minor issue concerns the way channels are displayed in overlapping mode. Previously, the used to be centered vertically, but now they stick to the top of the window. If it's easy to do, I'd prefer the former way (but if it's too complicated don't bother changing it).
8. The external link issue in the About dialog is fixed, but there is another issue: the version number is not displayed.

Since this PR keeps getting longer and longer, I suggest that we now focus on getting 0.6 ready and remove the color settings for now (please create a new issue/PR to remind us to work on this later on).

cbrnr commented:
@Yida-Lin, the About dialog is fixed - please pull in the latest commits of this branch before you continue working on other stuff.

cbrnr commented:
Regarding your comment about libxdf using standard library types and SigViewer using Qt types: First, loading a large >600MB XDF file takes <3s on my machine right now. So using that as an indicator, there's not much we need to change immediately. Second, have you tested this to see how much speedup you get? Third, can you point me to the relevant source code lines where this conversion takes place?

Yida-Lin commented:
@cbrnr Seriously? My laptop has almost all the newest processors and specs but it took me 6+ seconds to open the 600MB file :) Is it because it's windows? 

The copying below in `xdf_reader.cpp` is totally unnecessary in my opinion... 

            for (auto &row : stream.time_series)
            {
                ProgressBar::instance().increaseValue (1, progress_name);

                QSharedPointer<QVector<float32> > raw_data(new QVector<float32> (numberOfSamples, NAN));

                std::copy(row.begin(), row.end(), raw_data->begin() + startingPosition);
                QSharedPointer<DataBlock const> data_block(new FixedDataBlock(raw_data, XDFdata->majSR));

                channel_map_[channel_id] = data_block;
                channel_id++;

                std::vector<float> nothing;
                row.swap(nothing);
            }

Yida-Lin commented:
@cbrnr Let's see whether you like this design...

![image](https://cloud.githubusercontent.com/assets/20526646/23395299/e037536c-fd5b-11e6-9b52-c96dfe261d0f.png)

cbrnr commented:
@Yida-Lin, here's an up-to-date summary of what's left in this PR:

1. By default, SigViewer draws the channel separators (even though the option is unchecked). The lines are only drawn in the channel view, but not in the y-axis and label widgets. You can reproduce this issue if you delete SigViewer's settings (on Windows in the [registry](http://doc.qt.io/qt-5/qsettings.html#platform-specific-notes)).
2. Sometimes, the separators are not drawn - try changing the number of displayed channels. For me, the bottom separator is not drawn if I display just 1 channel.
3. The x-axis and y-axis scales are not synchronized with the window size. If you set the number of channels on the y-axis or the time range on the x-axis, these settings should stay constant even if you change the window size. Right now, this is not the case.
4. Another minor issue concerns the way channels are displayed in overlapping mode. Previously, the used to be centered vertically, but now they stick to the top of the window. If it's easy to do, I'd prefer the former way (but if it's too complicated don't bother changing it).

Yida-Lin commented:
@cbrnr Hi Clemens,

I made separate commits for these problems respectively.

> By default, SigViewer draws the channel separators (even though the option is unchecked). 

Fixed. 

> Sometimes, the separators are not drawn - try changing the number of displayed channels. For me, the bottom separator is not drawn if I display just 1 channel.

Weird. I haven't figured it out yet.

> The x-axis and y-axis scales are not synchronized with the window size. 

To me the y_axis is always synced well but x is not. I will see what I can do.

> Another minor issue concerns the way channels are displayed in overlapping mode. Previously, the used to be centered vertically, but now they stick to the top of the window. 

I can't help much on this issue because if I revert this, the x widget problem reappears. I haven't found a way that can suffice both ways.

> The status bar is not in sync with the menu entry in the View menu. I think this issue occurs if you start fresh, i.e. delete the settings and then toggle the status bar in the View menu and observe the check.

Fixed.

cbrnr commented:
Simplified `main.cpp`, should fix #40.

cbrnr commented:
Your attempt to fix issue 3 goes into the right direction! IMO, the issue is solved for the x-axis, because resizing the window horizontally does not change the amount of time displayed. There are two minor issues though: (1) The default time duration is now only 0.1s - this needs to be changed back to 5s (or 10s). (2) I get the following runtime message:

    QObject::connect: No such slot sigviewer::SignalBrowserView::selfUpdatingFinished()

I don't think that it is already working for the y-axis, but almost. It seems there is some inaccuracy in a relevant calculation, because try this: start with a specific (small) window size:

![screen shot 2017-03-01 at 12 03 07](https://cloud.githubusercontent.com/assets/4377312/23457169/138aedca-fe77-11e6-98bd-3f3d5342e2da.png)

Then, resize the window vertically several times. You will notice that the space below the last channel magically increases when you increase the window size, and if you decrease the window size this white space does not shrink:

![screen shot 2017-03-01 at 12 04 45](https://cloud.githubusercontent.com/assets/4377312/23457211/452dd00e-fe77-11e6-8165-dffffe62bddf.png)

Finally, a new issue. For some reason, both sliders are rendered as vertical sliders by default:

![screen shot 2017-03-01 at 12 33 23](https://cloud.githubusercontent.com/assets/4377312/23458163/4fa3ff32-fe7b-11e6-877d-505a6df36fa7.png)

This happens when the settings are cleared. When I enlarge the window and restart SigViewer, they look normal.

cbrnr commented:
@Yida-Lin can we keep my last commit please - I forgot to push some changes, will do that when I'm in my office. I tested it on macOS and Linux and the settings were working. If fullscreen isn't remembered, I know what setting to add.

Yida-Lin commented:
@cbrnr Sure Clemens. I have already reversed the changes.

I have different opinion on status bar: to me this is a very crucial and straight forward way to see the scope of the file. I propose to set default as shown.

The x axis sync problem is not solved. If I set the initial value of the sec per page spinbox to anything other than 0.1, it crashes. I don't have a clear idea about it. Also, setting to any arbitrary number is not a good idea: the file length varies significantly. 

 This 'fun game' exists earlier in x widget toggling as well: until I set the central widget to alignTop. However, I tried multiple methods, none of them worked so far.

Yida-Lin commented:
@cbrnr Hi Clemens,

There are mistakes in those several commits. SigViewer won't even be able to start, as shown in the figure: 

![image](https://cloud.githubusercontent.com/assets/20526646/23538357/90f070f8-ffa0-11e6-86f1-783249089983.png)

You can reproduce the issue if you clean your solution, and rebuild it.

I created a temporary branch for myself and rolled back to commits before, then it worked.

Yida-Lin commented:
@cbrnr Actually, surprisingly I have found a solution to solve the x widget toggling problem while keep the central widget to alignVCenter. Also, it solves the 'fun game' when switching mode as well.

I will push this commit once the errors in the previous commits are fixed.

cbrnr commented:
@Yida-Lin please feel free to revert all commits that introduced errors!

Yida-Lin commented:
@cbrnr Thank you Clemens. Please see my new improvements.

Now no matter how you spam Ctrl + num or toggling x widget, the main widget will stay robust :) 

Also, the overlapping will center around vertical center. 

The newly added color check box is just a concept: can be easily taken out if needed.

cbrnr commented:
Great! I think we should finish this up now. Let's summarize what remains to be done and what we will move to separate PRs:

1. The default time setting for XDF files is 0.1s (for GDF it seems to be 3.8s) - can you change this value for all file types to 5s?
2. The default value for channels seems to be 3 for GDF (don't know about XDF). I think it would be good to either show all channels by default (but at most e.g. 32 or some other number). What do you think?
3. Disable the color settings for now.

Solve in separate PRs:
1. The displayed time range and channel number should remain fixed even when resizing the window.
2. Unify/simplify QSettings.
3. Clean up/simplify main.cpp and main_window.cpp

Anything else?

Yida-Lin commented:
@cbrnr Cool!

1. I don't think it's 0.1s for XDF. It varies depending on files. Changing it to 5s is temporarily impossible because it introduces a crashing bug (for setting the spinbox to arbitrary value other than 0.1). Also, because file length and sample rates varies significantly, setting the starting length to a scalar doesn't sound like appropriate.

![image](https://cloud.githubusercontent.com/assets/20526646/23542961/f02bd7de-ffbd-11e6-8f1c-e3b5d9342498.png)


2. Also doesn't seem to be 3 or any number to me. It's automatically done (just as before). I think we should we should just keep the way it is.

![image](https://cloud.githubusercontent.com/assets/20526646/23542967/f8ab0f24-ffbd-11e6-9710-21cde8a7209d.png)

3. Will do.

I think we should set a check for update function, which doesn't force user to update. Of course, we can do this in a separate PR.

cbrnr commented:
OK. Seems like everything works fine (or at least the same as before). Please let me know when you feel that this PR can be merged. Also, please collect all changes that we discarded here in separate issues just so we won't forget them. Please also add new ideas (such as the check for update feature) as new issues! Then, we can create PRs to address each issue later on.

Yida-Lin commented:
@cbrnr Hi Clemens! I think for now it's ready, excluding those issues we mentioned above. I will make new issues on them separately tomorrow.

cbrnr commented:
Thanks @Yida-Lin!

Yida-Lin commented:
@cbrnr  I thank you Clemens! I really enjoyed working on it :)
